component/status/status.go

```diff
@@ -0,0 +1,194 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status // import "go.opentelemetry.io/collector/component/status"
+
+import (
+	"errors"
+	"time"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+// EventType represents an enumeration of status event types
+type EventType int
+
+const (
+	// OK indicates the producer of a status event is functioning normally
+	OK EventType = iota
```

Comment:
On the prototype @pmm-sumo pointed me towards this document: https://docs.google.com/document/d/15nJTAzomzUJvwhINvTxHuaxUNe7dfgfG3gS1Tib8TGM/edit#. This screenshot is taken from it:
<img width="649" alt="Screen Shot 2022-05-16 at 5 54 02 PM" src="https://user-images.githubusercontent.com/2513372/168705808-4c337c42-fd93-45bb-a09e-d183d6c8e1d5.png">

It shows a component reporting a stream of status events. I think the conventions need to be better defined, and I could use some help from the author of that document (if anyone knows who it is) to solidify them. My initial thought is that a component should report its status on startup, and if they do work on a periodic basis, they should report the result of that work (an export result for an exporter, or the result of a scrape for a scraping receiver, etc). I'm open to suggestions on all of this.

Reply:
@mwear that would be me :) 

The document was describing ideas on better own telemetry reporting, largely with early work on [remote management](https://github.com/open-telemetry/opamp-spec) in mind. 

Health event reporting I mentioned there is somewhat based on the [Health Events](https://help.sumologic.com/Manage/Health_Events#health-event-log-example) capability at Sumo. At a given time, each component can have _healthy_ or _unhealthy_ status which could be reported upstream. To avoid state flapping, there's the concept of having some sort of threshold before the state changes. Having said that, that was just an idea (and it wasn't included in OpAMP spec). I think the most important thing now is to define the internal API for status reporting (which can then be used for generating health events like described in the docs using status listeners and such).

Let me know if you need any more details on that (also feel free to ping me on CNCF Slack to make it more synchronous, my display name there is "Przemek Maciolek")